### ISSUE TITLE
disable nullable behavior globally

### ISSUE THREAD

**Issue by dryajov** [artifacthealth/hydrate-mongodb#15]
Currently, `null` fields and `undefined` fields are treated the same, and are automatically removed from the collection on each flush. IMO this is a very bad default because it implicitly modifies existing data as a side effect, which is both unexpected and not needed by default in most cases, and its specially bad when working with existing data. 

Is there any way of disabling the behavior globally without having to annotate each field as nullable? If there isn't, I would suggest that the default is reverted to not alter any data unless explicitly enabled, and furthermore add a way of globally controlling this functionality.

**Comment by dryajov** [comment]
PR #16 addresses this.

**Comment by meirgottlieb** [comment]
Hey @dryajov, 

Thank you for the suggestion. I agree there should be a way to change this globally without having to annotate every field. However, I do not feel this should be the default behavior. 

I took a look at the pull request. The problem with that approach is that it only works for fields annotated with Field, which is not a required annotation. It is only required that every field have some sort of annotation so that type information is added by TypeScript. For example, if you have an array and you are already adding the ElementType annotation, there is no need to add the Field annotation. 

I'll have to think some more about the best approach to add this configuration item.

As far as how to test annotations, take a look at [annotationMappingProvider.tests.ts](https://github.com/artifacthealth/hydrate-mongodb/blob/master/tests/mapping/providers/annotationMappingProvider.tests.ts).

**Comment by dryajov** [comment]
I'm either way on the default. Tho I think my suggestion is a more sane default rather than writing to the DB implicitly, there are many reasons for that, one that comes to mind is performance - this is additional load that is not visible untill you hit prod, for massive applications this would be a nasty surprise. 

As for the check being in the field, I think it would be sane to move it to where null/undefined fields get removed, and before the dirty check is made.

**Comment by meirgottlieb** [comment]
I hear you. Although it's just an issue when trying to map legacy data. Let me think about it.

**Comment by dryajov** [comment]
Absolutely, but this would be a reason not to use it with existing datastores, which could potentially be a large portion of the users. 

Anyways, thanks for getting back so quickly. This is something that we want to ship in the next couple of weeks, you don't have to rush anything obviously - we can always fork it for now, but if possible it be great to get this in ASAP.

**Comment by meirgottlieb** [comment]
What about `undefined`? MongoDB can store a value of `undefined` which is different than not having a value at all. For example, here is output from the mongo shell:

```
> db.test.insert({ a: null, b: undefined })
WriteResult({ "nInserted" : 1 })
> db.test.find({ b: { $exists: 1}})
{ "_id" : ObjectId("5969317d72ccf76744a93dde"), "a" : null, "b" : undefined }
```

**Comment by dryajov** [comment]
Good question. I think the behavior should be to leave the fields unchanged unless told otherwise. Again, for new datasources, this probably makes little difference, old datasource might attach some application level semantics to those fields. So if `nullable` is `true`, leave both `null` and `undefined` as is, if `false`, then remove them.

**Comment by meirgottlieb** [comment]
What about fields in a document that are not represented in the mapping? Currently they would be dropped.

**Comment by dryajov** [comment]
Why would they? The framework should not make assumptions about the underlying datastore at all, I think those fields should still exist in the retrieved document, but ignored during the dirty check.

**Comment by meirgottlieb** [comment]
They would be ignored during a dirty check. But if something changed in the Entity they would be dropped from the serialized document.

**Comment by dryajov** [comment]
I think not mapped fields should be wrapped with some sort of ignore metadata, so that they are skipped during dirty checking, i.e. my application doesn't care about them, but other applications might, so dropping them makes no sense, in fact its dangerous and unexpected behavior all together, at the very least this things should be bolded out in the docs.

**Comment by dryajov** [comment]
This can be handled with getters/setters, if a field that is not mapped is tried to be altered and exception should be thrown.

**Comment by meirgottlieb** [comment]
Looks like persisting `undefined` is not an option. The underlying driver either converts `undefined` to `null` or drops them depending on an option passed to `insert`. See https://github.com/christkv/mongodb-core/issues/31.

**Comment by dryajov** [comment]
I guess thats not such a big issue - in this case we should probably be consistent with the driver. IMO the fact that mongo allows storing `undefined` is a pretty weird quirk 🏸

**Comment by dryajov** [comment]
this comment - https://github.com/christkv/mongodb-core/issues/31#issuecomment-124011112 is a good behavior to emulate, IMO.

**Comment by dryajov** [comment]
@meirgottlieb 

I think we can add the check for the global `nullable` here - https://github.com/artifacthealth/hydrate-mongodb/blob/167bdfaa1baa6099bb6d98822fbb57952ebb6808/src/mapping/objectMapping.ts#L152. The only issue I see is that there doesn't seem to be a way of getting at the config object right now.

Also, seems like there is already a place were the missing properties are dropped - https://github.com/artifacthealth/hydrate-mongodb/blob/167bdfaa1baa6099bb6d98822fbb57952ebb6808/src/mapping/objectMapping.ts#L175, maybe change the default to not drop them there as well?

Thoughts?

**Comment by meirgottlieb** [comment]
I'm leaning toward making the null change and making it the default behavior. I have a pretty good idea of where to do it but will have to make a change to how the flag is represented on the property. 

As far as dropping unmapped properties - Trying to copy all unmapped fields into the serialized document could be very expensive and probably not the right approach. Perhaps using $set to update a document could work instead of doing a document replacement (which is what I do now). I've considered doing a diff to only update properties that have changed and then using $set/$unset but figured build the diff may be more expensive then just replacing the whole document. 

With regard to the second reference above - That's a different situation and is expected behavior. That is for when `refresh` is called to refresh the state of an entity from the database. That is saying that if the entity in memory has a value and the document in the database does not, it drops the field  on the entity in memory when `refresh` is called.

**Comment by dryajov** [comment]
>I'm leaning toward making the null change and making it the default behavior. I have a pretty good idea of where to do it but will have to make a change to how the flag is represented on the property.

👍 

>As far as dropping unmapped properties - Trying to copy all unmapped fields into the serialized document could be very expensive and probably not the right approach. Perhaps using $set to update a document could work instead of doing a document replacement (which is what I do now). I've considered doing a diff to only update properties that have changed and then using $set/$unset but figured build the diff may be more expensive then just replacing the whole document.

I agree, using `$set`/`$unset` is probably the sane thing to do here, that way only the parts of the document that require changes are touched. On the other hand, maybe a full merge is not required?
It would seem that when the entity is created from the retrieved document it should be possible to simply add those non mapped field to the entity but mark them as internal/ignored or some such, that way when serialization happens, nothing else needs to be copied? Perhaps you can point me to the relevant places in the code, so that I get a better idea of the limitations, and perhaps provide more on point suggestions :)

>With regard to the second reference above - That's a different situation and is expected behavior. That is for when refresh is called to refresh the state of an entity from the database. That is saying that if the entity in memory has a value and the document in the database does not, it drops the field on the entity in memory when refresh is called.

👍

**Comment by meirgottlieb** [comment]
I don't think copying unmapped fields into the entity is a good idea. You could have unexpected results such as overwriting private class properties. The original document is actually kept around in the session for dirty checking. If anything, I would copy all the unmapped field back to the document when serializing. But using $set/$unset is probably the best approach.

**Comment by dryajov** [comment]
Do $set/$unset support applying partial documents without having to use the clunky dot ('.') notation? Otherwise, copying from the original is probably the best approach.

**Comment by meirgottlieb** [comment]
You have to use the dot notation. Applying partial changes to embedded arrays is very limited as well, which is one of the reasons I abandoned that approach originally.

**Comment by dryajov** [comment]
ah, that sounds cumbersome... maybe just copying from the original doc... or the other way around, copying from the entity onto the original doc, if anything the original doc is probably going to be larger?

**Comment by meirgottlieb** [comment]
Hey @dryajov, I published an updated version to npm that includes the `nullable` option in `Configuration`. I decided to leave the default as `false` to not break existing behavior. Instead, there should be a discussion in the README for considerations when mapping legacy data.

I'm going to go ahead and close out this ticket. If you want to pursue an option of preserving unmapped fields in legacy data, feel free to open another ticket. I don't have time to work on that now but I can point you in the right direction if you want to tackle it.
